### PR TITLE
[ML] Evolve use of versions in decision making

### PR DESCRIPTION
**Pull request by edsavage**
In this first pass at evolving the way the ML and Transform plugins use version numbers, where possible TransportVersions are used to answer questions like "is the cluster in such a state to support a particular feature".

Labelling as a `>non-issue` as this _should_ have no outwardly discernible effects.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java [hunk 1]
```diff
@@ -186,11 +187,11 @@ protected void masterOperation(
...
             }
 
-            Version minCompatibilityVersion = config.getModelDefinition().getTrainedModel().getMinimalCompatibilityVersion();
-            if (state.nodes().getMinNodeVersion().before(minCompatibilityVersion)) {
+            TransportVersion minCompatibilityVersion = config.getModelDefinition().getTrainedModel().getMinimalCompatibilityVersion();
+            if (state.getMinTransportVersion().before(minCompatibilityVersion)) {
                 listener.onFailure(
                     ExceptionsHelper.badRequestException(
-                        "Definition for [{}] requires that all nodes are at least version [{}]",
+                        "Definition for [{}] requires that all transport versions are at least [{}]",
```
Comment (hendrikmuhs) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java:194]: Nit: This is a very technical message, I wonder if we can make this more consumable, saying that the versions are different.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportUpgradeTransformsAction.java [hunk 2]
```diff
@@ -95,12 +96,12 @@ protected void masterOperation(Task ignoredTask, Request request, ClusterState s
         TransformNodes.warnIfNoTransformNodes(state);
 
         // do not allow in mixed clusters
-        if (state.nodes().getMaxNodeVersion().after(state.nodes().getMinNodeVersion())) {
+        if (TransportVersionUtils.isMinTransformVersionSameAsCurrent(state) == false) {
             listener.onFailure(
                 new ElasticsearchStatusException(
-                    "Cannot upgrade transforms. All nodes must be the same version [{}]",
+                    "Cannot upgrade transforms. All transport versions must be the same [{}]",
```
Comment (hendrikmuhs) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportUpgradeTransformsAction.java:102]: nit, similar suggestion:

"Cannot upgrade transform while cluster upgrade is in progress".

**Review by hendrikmuhs — COMMENTED** [review]
LGTM, I wonder if we can improve messages

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/TransportVersionUtils.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    public static TransportVersion getMinTransportVersion(ClusterState state) {
+        return state.getMinTransportVersion();
+    }
+
+    public static TransportVersion getCurrentTransportVersion() {
+        return TransportVersion.current();
+    }
+
+    public static boolean isMinTransformVersionSameAsCurrent(ClusterState state) {
```
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/TransportVersionUtils.java:24]:
```suggestion
    public static boolean isMinTransportVersionSameAsCurrent(ClusterState state) {
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java [hunk 4]
```diff
@@ -186,11 +187,11 @@ protected void masterOperation(
...
             }
 
-            Version minCompatibilityVersion = config.getModelDefinition().getTrainedModel().getMinimalCompatibilityVersion();
-            if (state.nodes().getMinNodeVersion().before(minCompatibilityVersion)) {
+            TransportVersion minCompatibilityVersion = config.getModelDefinition().getTrainedModel().getMinimalCompatibilityVersion();
+            if (state.getMinTransportVersion().before(minCompatibilityVersion)) {
                 listener.onFailure(
                     ExceptionsHelper.badRequestException(
-                        "Definition for [{}] requires that all nodes are at least version [{}]",
+                        "Cannot create model [{}] while cluster upgrade is in progress. All transport versions must be at least [{}]",
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java:194]: These are user-facing error messages, and they won't have a clue what a transport version is, so we shouldn't mention it.

```suggestion
                        "Cannot create model [{}] while cluster upgrade is in progress.",
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java [hunk 5]
```diff
@@ -136,12 +137,13 @@ protected void masterOperation(
             return;
         }
 
-        if (state.nodes().getMaxNodeVersion().after(state.nodes().getMinNodeVersion())) {
+        if (TransportVersionUtils.isMinTransformVersionSameAsCurrent(state) == false) {
             listener.onFailure(
                 new ElasticsearchStatusException(
-                    "Cannot start a new model deployment as not all nodes are on version {}. All nodes must be the same version",
+                    "Cannot start a new model deployment as not all transport versions used by the cluster are on version {}. "
+                        + "All transport versions must be the same version.",
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java:144]:
```suggestion
                    "Cannot start model [{}] while cluster upgrade is in progress.",
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java [hunk 6]
```diff
@@ -109,13 +110,14 @@ protected void masterOperation(Task task, Request request, ClusterState state, A
             return;
         }
 
-        if (state.nodes().getMaxNodeVersion().after(state.nodes().getMinNodeVersion())) {
+        if (TransportVersionUtils.isMinTransformVersionSameAsCurrent(state) == false) {
             listener.onFailure(
                 ExceptionsHelper.conflictStatusException(
-                    "Cannot upgrade job [{}] snapshot [{}] as not all nodes are on version {}. All nodes must be the same version",
+                    "Cannot upgrade job [{}] snapshot [{}] as not all transport versions are {}."
+                        + " All transport versions must be the same version",
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java:117]:
```suggestion
                    "Cannot upgrade job [{}] snapshot [{}] while cluster upgrade is in progress.",
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java [hunk 7]
```diff
@@ -589,13 +589,14 @@ private void updateNumberOfAllocations(
             );
             return;
         }
-        if (clusterState.nodes().getMinNodeVersion().before(DISTRIBUTED_MODEL_ALLOCATION_VERSION)) {
+        if (clusterState.getMinTransportVersion().before(DISTRIBUTED_MODEL_ALLOCATION_TRANSPORT_VERSION)) {
             listener.onFailure(
                 new ElasticsearchStatusException(
-                    "cannot update number_of_allocations for deployment with model id [{}] while there are nodes older than version [{}]",
+                    "cannot update number_of_allocations for deployment with model id [{}] while there are transport"
+                        + " versions older than [{}]",
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java:596]:
```suggestion
                    "cannot update number_of_allocations for deployment with model id [{}] while cluster upgrade is in progress.",
```

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportUpgradeTransformsAction.java [hunk 8]
```diff
@@ -95,12 +96,12 @@ protected void masterOperation(Task ignoredTask, Request request, ClusterState s
         TransformNodes.warnIfNoTransformNodes(state);
 
         // do not allow in mixed clusters
-        if (state.nodes().getMaxNodeVersion().after(state.nodes().getMinNodeVersion())) {
+        if (TransportVersionUtils.isMinTransformVersionSameAsCurrent(state) == false) {
             listener.onFailure(
                 new ElasticsearchStatusException(
-                    "Cannot upgrade transforms. All nodes must be the same version [{}]",
+                    "Cannot upgrade transforms while cluster upgrade is in progress. All transport versions must be the same [{}]",
```
Comment (droberts195) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportUpgradeTransformsAction.java:102]:
```suggestion
                    "Cannot upgrade transforms while cluster upgrade is in progress.",
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java [hunk 9]
```diff
@@ -186,11 +187,11 @@ protected void masterOperation(
...
-            Version minCompatibilityVersion = config.getModelDefinition().getTrainedModel().getMinimalCompatibilityVersion();
-            if (state.nodes().getMinNodeVersion().before(minCompatibilityVersion)) {
+            TransportVersion minCompatibilityVersion = config.getModelDefinition().getTrainedModel().getMinimalCompatibilityVersion();
+            if (state.getMinTransportVersion().before(minCompatibilityVersion)) {
                 listener.onFailure(
                     ExceptionsHelper.badRequestException(
-                        "Definition for [{}] requires that all nodes are at least version [{}]",
+                        "Cannot create model [{}] while cluster upgrade is in progress.",
                         config.getModelId(),
                         minCompatibilityVersion.toString()
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java:196]: This argument is redundant now after the change to the message. The same problem will exist for the other messages that were changed.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java [hunk 10]
```diff
@@ -136,13 +137,9 @@ protected void masterOperation(
...
 
-        if (state.nodes().getMaxNodeVersion().after(state.nodes().getMinNodeVersion())) {
+        if (TransportVersionUtils.isMinTransportVersionSameAsCurrent(state) == false) {
             listener.onFailure(
-                new ElasticsearchStatusException(
-                    "Cannot start a new model deployment as not all nodes are on version {}. All nodes must be the same version",
-                    RestStatus.FORBIDDEN,
-                    state.getNodes().getMaxNodeVersion()
-                )
+                new ElasticsearchStatusException("Cannot start model [{}] while cluster upgrade is in progress.", RestStatus.FORBIDDEN)
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java:142]:
```suggestion
                new ElasticsearchStatusException("Cannot start model deployment [{}] while cluster upgrade is in progress.", request.getDeploymentId(), RestStatus.FORBIDDEN)
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java [hunk 11]
```diff
@@ -276,14 +276,14 @@ public void createNewModelAssignment(
         StartTrainedModelDeploymentAction.TaskParams params,
         ActionListener<TrainedModelAssignment> listener
     ) {
-        if (clusterService.state().nodes().getMinNodeVersion().before(DISTRIBUTED_MODEL_ALLOCATION_VERSION)) {
+        if (clusterService.state().getMinTransportVersion().before(DISTRIBUTED_MODEL_ALLOCATION_TRANSPORT_VERSION)) {
             listener.onFailure(
                 new ElasticsearchStatusException(
-                    "cannot create new assignment [{}] for model [{}] while there are nodes older than version [{}]",
+                    "cannot create new assignment [{}] for model [{}] while there are transport versions older than [{}]",
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java:282]:
```suggestion
                    "cannot create new assignment [{}] for model [{}] while cluster upgrade is in progress",
```

**Review by droberts195 — APPROVED** [review]
LGTM

**Pull request merged**